### ISSUE TITLE
Environment.FailFast() can be used to crash the container

### ISSUE THREAD

**Issue by lennartb-** [ashmind/SharpLab#1204]
Hi, for my master's thesis I'm currently researching how (untrusted) C# script code can be run safely and isolated, without abilty to compromise the system it is executed on. Part of that is a bit of trying around how other tools (like SharpLab or .NET Fiddle) behave with potentially malicious code.

I ran

```
using System;
public class C {
    public static void Main() {
        Environment.FailFast("test");
    }
}
```

... which results in `(Container crashed, next container will be available in 15 seconds)`.

I wasn't sure if that is intended/okay behavior, so I wanted to let you to know about it. I also tried stuff like `Directory.EnumerateFiles` (which was already covered in #1163), and some other members in the `Environment` class, which leaks the current working directory and command line, but I'm not sure if that's really a secret.

**Comment by ashmind** [comment]
Thanks for reporting! It is very interesting topic I can talk a lot about.
I'll write a bit in case, hopefully in a way that could be useful for you research.

### Attack Types

When thinking on this topic, I have considered several attack vectors:
1. Exiting or otherwise affecting the host (e.g. `FailFast`)
2. Leaking information from the host (e.g. files, environment veriables)
3. Overflowing the stack
4. Over-allocating memory
5. Overusing host CPU

### Unbreakable

Previous versions of SharpLab ran code in-process. Those versions had a rather strict manual validation of things you can do, as  you [can still see it in .NET Framework version](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYwCYgNQB8ACATAAwCwAUEQIwUUCiqAbgJaID2qAtgKapwB0AMQCGzADYiAznAAUAIjjdpcgJQBuIA===).

This was based on a custom library called [Unbreakable](https://github.com/ashmind/Unbreakable). Unbreakable provided the following protections:
* Specific list of safe methods to use (`Environment.FailFast` not on that list)
* Estimation of stack size and limit on total estimated size
* Estimation of allocated memory, and limit on total estimated memory
* Limit on execution time

This was reasonably comprehensive, and worked relatively well, but was a maintenance nightmare:
* With each new .NET version, I needed to keep manually reviewing and adding safe APIs to the list
* It was very hard to properly estimate allocated memory just from methods called
* People asked for an unsafe mode, and I wanted to add an IL mode as well, and with pointers every analysis was even harder

So in the end, I decided to give up on in-process model.

### Fragile

Current version of SharpLab uses a lightweight sandbox model based on Windows AppContainer Isolation. This is encapsulated in a library called [Fragile](https://github.com/ashmind/SharpLab/tree/main/source/%23external/Fragile/Fragile), which is not published separately at the moment. What it does is very similar to a Chrome browser sandbox. Or you can think of it as a light-weight version of Docker.

Basically, instead of running the code in the main process, it runs the code in a separate process with hugely reduced permissions, enforced by the OS.

Here is how it interacts with the attack types above:
**1. Exiting or otherwise affecting the host (e.g. `FailFast`)**

This only exists the container. Since your container is personal, you are only creating a delay for yourself.
This can be used for a DoS attack if you do it fast enough in parallel, but SharpLab is not immune to dedicated DoS attacks either way, and aggressive DoS was not in my scope.

**2. Leaking information from the host (e.g. files, environment variables)**

The sandbox prevents this -- as you can see, you have only limited access to environment or file system. You have full access to the memory of the sandboxed process, but it's your personal process, so you can't get access to other user sessions.

**3. Overflowing the stack**

This only affects the container.

**4. Over-allocating memory**
**5. Overusing host CPU**

These are controlled by the Windows Job Object limits -- again, enforced by the operating system.

### Proper Alternatives

A more enterprise approach I would take, if I was doing this at work, would be to look at something like Azure Container Instances or Azure Kubernetes, where the container does not need a custom library, and is managed by Microsoft. Since cloud vendors already worry about malicious containers affecting others, they have rather comprehensive security in place to prevent this, which we don't have to reinvent.

The only reason I haven't done this for SharpLab yet is because I am not fully sure on the costs compared to my current solution, and as far as I remember setting the limits was a bit harder (which would also contribute to unexpected costs if exploited).

However if I had a proper team working on it, I would definitely start with this, rather than all the custom choices above.

**Comment by lennartb-** [comment]
Thanks for your answer! This fills some gaps I had and reinforced a lot of the conclusions I made.

AppContainers are something I already implemented for testing, but the incredible lack of documentation on MSDN made me question whether these are (still) actively used. I basically poached my implementation from various blog posts and samples on GitHub. Good to know these are still a valid way to do it. I knew Windows Store/UWP Apps and Chrome used them as part of their sandbox, but most of the resources I found were pretty old, and I wasn't sure whether this was still true for today. I found them pretty hard to use and difficult to get “right”, especially for potentially business-critical applications.

Full-on containerization (I examined Docker and OS-level virtualization in general in my thesis) seems more suited for my use case (isolation of applications build from code-generators) anyway. Preconfiguring a container image and recompiling the application is pretty easy to do with them, and with their popularity in recent years, there's a lot of tooling, obviously.

Allow/deny lists like with Unbreakable seem like an initial good choice from a business perspective (“We just forbid our users to use reflection/IO/processes!”), but as you said, it's hard to maintain and can be little more than a speed bump. I also mentioned CAS and AppDomains, but both are managed-only, deprecated and ultimately not as secure as containerization anyway.

**Comment by yyjdelete** [comment]
If it infact use sandbox now for .NetCore and doesn't limit api, should the limit `Type XXX has a static constructor, which is not supported by SharpLab JIT decompiler.` also be relax for `JIT ASM`?

**Comment by ashmind** [comment]
@yyjdelete That's a good question. The sandbox is not currently used for JIT ASM since JIT ASM mostly does not run code, with the exception of static constructors and module initializers. However I do plan to eventually move to a different approach for ASM, e.g. would be great to decompile to ARM assembly without having to run the whole site on ARM.

**Comment by ashmind** [comment]
Closing as this one should be fine, however I very much appreciate any further security reports.
